sourcehut: init

eadwu opened the pull request:
###### Motivation for this change
Totally didn't steal `pgpy` and `sqlalchemy-utils` from #50328.

###### Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- [x] Tested using sandboxing ([nix.useSandbox](http://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](http://nixos.org/nix/manual/#sec-conf-file) on non-NixOS)
- Built on platform(s)
   - [x] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nox --run "nox-review wip"`
- [ ] Tested execution of all binary files (usually in `./result/bin/`)
- [ ] Determined the impact on package closure size (by running `nix path-info -S` before and after)
- [ ] Assured whether relevant documentation is up to date
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

---

dotlambda commented on pkgs/development/python-modules/weasyprint/default.nix:
```diff
@@ -0,0 +1,48 @@
+, cffi, pyphen, cairosvg, html5lib, tinycss2, cairocffi, cssselect2
+, pytestcov, pytest-flake8, pytest-isort }:
+
+# python3 only
```
`disabled = !isPy3k`

dotlambda commented on pkgs/development/python-modules/weasyprint/default.nix:
```diff
@@ -0,0 +1,48 @@
+  ];
+
+  meta = with stdenv.lib; {
+    homepage = "https://weasyprint.org/";
```
```suggestion
    homepage = https://weasyprint.org/;
```

dotlambda commented on pkgs/development/python-modules/sshpubkeys/default.nix:
```diff
@@ -0,0 +1,30 @@
+  doCheck = false;
+
+  meta = with stdenv.lib; {
+    homepage = "https://github.com/ojarva/python-sshpubkeys";
```
```suggestion
    homepage = https://github.com/ojarva/python-sshpubkeys;
```

dotlambda commented on pkgs/development/python-modules/sshpubkeys/default.nix:
```diff
@@ -0,0 +1,30 @@
+  ];
+
+  # ModuleNotFoundError: No module named 'tests'
+  doCheck = false;
```
You might have to add `PYTHONPATH=".:$PYTHONPATH"`.

eadwu replied:
```
======================================================================
ERROR: tests (unittest.loader._FailedTest)
----------------------------------------------------------------------
ImportError: Failed to import test module: tests
Traceback (most recent call last):
  File "/nix/store/ydk0mfpvn9smcmn72wc9i20slv1d2b79-python3-3.7.2/lib/python3.7/unittest/loader.py", line 154, in loadTestsFromName
    module = __import__(module_name)
ModuleNotFoundError: No module named 'tests'
```

eadwu replied:
Setting explicitly in the derivation, `PYTHONPATH = ".:$PYTHONPATH";` errors with 
```
Found duplicated packages in closure for dependency 'sshpubkeys':
  sshpubkeys 3.1.0 (/nix/store/2vd539zsgavw0c7vajsy4a18wwp6hcsi-python2.7-sshpubkeys-3.1.0/lib/python2.7/site-packages)
  sshpubkeys 3.1.0 (/build/sshpubkeys-3.1.0)
```

eadwu replied:
Tried including that PR as a patch but didn't seem to work, just changed it to use the Github source since the `tests/` directory is there.

dotlambda replied:
Please include a comment with reason for not using `fetchPyPI`.

dotlambda commented on pkgs/development/python-modules/weasyprint/default.nix:
```diff
@@ -0,0 +1,48 @@
+  doCheck = false;
+  checkInputs = [
+    pytestcov
+    pytestrunner
```
That's a duplicate.

dotlambda commented on pkgs/development/python-modules/sqlalchemy-utils/default.nix:
```diff
@@ -0,0 +1,39 @@
+    sqlalchemy
+  ];
+
+  doCheck = false;
```
Why?

dotlambda replied:
The tests are accessing the network (localhost). That's why they're failing. And if that weren't prohibited by the sandbox, there still would be no database to access.

dotlambda commented on pkgs/development/python-modules/weasyprint/default.nix:
```diff
@@ -0,0 +1,48 @@
+{ stdenv, fetchPypi, buildPythonPackage
```
```suggestion
{ lib, fetchPypi, buildPythonPackage
```

dotlambda commented on pkgs/development/python-modules/weasyprint/default.nix:
```diff
@@ -0,0 +1,48 @@
+    pytest-isort
+  ];
+
+  meta = with stdenv.lib; {
```
```suggestion
  meta = with lib; {
```
Please also do so for the other packages.

dotlambda commented on pkgs/development/python-modules/pytest/default.nix:
```diff
@@ -23,6 +23,8 @@ let generic = { version, sha256 }:
       ++ stdenv.lib.optionals (pythonOlder "3.6") [ pathlib2 ];
 
+    # https://hydra.nixos.org/build/87478471/nixlog/1
+    doCheck = false;
```
That's for `pytest_32` used by `celery`. This has been fixed in fae1f7afc144ba11bc8fa340878b85956291280b, which will be merged with https://github.com/NixOS/nixpkgs/pull/54306.

dotlambda commented on pkgs/development/python-modules/pgpy/default.nix:
```diff
@@ -0,0 +1,34 @@
+
+  meta = with stdenv.lib; {
+    homepage = https://github.com/SecurityInnovation/PGPy;
+    description = "Pretty Good Privacy for Python 2 and 3.";
```
No period: https://nixos.org/nixpkgs/manual/#sec-standard-meta-attributes

dotlambda commented on pkgs/development/python-modules/celery/default.nix:
```diff
@@ -29,6 +29,8 @@ in buildPythonPackage rec {
   };
 
+  # SyntaxError: invalid syntax
+  doCheck = false;
```
A syntax error cannot be fixed by disabling tests. This is probably because celery does not support Python 3.7 yet: 932f0baec20c6b0947605115f78240d85ed926b4.

dotlambda commented on pkgs/development/python-modules/atpublic/default.nix:
```diff
@@ -0,0 +1,29 @@
+
+  meta = with stdenv.lib; {
+    homepage = "https://public.readthedocs.io/en/latest/";
+    description = "@public for populating __all__";
```
missing

eadwu replied:
Missing?

dotlambda replied:
Sorry. I though that was some kind of placeholder. It is obviously not but needs to be changed to a real description.

eadwu replied:
Ah okay, I just copied the description from `setup.py` which was `public -- @public for populating __all__`.

dotlambda commented on pkgs/development/python-modules/atpublic/default.nix:
```diff
@@ -0,0 +1,29 @@
+  doCheck = false;
+
+  meta = with stdenv.lib; {
+    homepage = "https://public.readthedocs.io/en/latest/";
```
```suggestion
    homepage = https://public.readthedocs.io/en/latest/;
```

dotlambda commented on pkgs/development/python-modules/atpublic/default.nix:
```diff
@@ -0,0 +1,29 @@
+      __all__ and optionally the module globals.
+
+      This provides both a pure-Python implementation and a C implementation.  It is
+      proposed that the C implementation be added to builtins_ for Python 3.6.
```
Has the proposal been adopted?

eadwu replied:
It looks like it might've been rejected? https://bugs.python.org/issue26632

dotlambda commented on pkgs/development/python-modules/atpublic/default.nix:
```diff
@@ -0,0 +1,29 @@
+  };
+
+  # TypeError: calling <function public at 0x7ffff26dbb70> returned None, not a test
+  doCheck = false;
```
That should be fixed.

eadwu replied:
Don't know if this is how it should be done but running `pytest --pyargs public` in `checkPhase` fixes it.

dotlambda commented on pkgs/development/python-modules/atpublic/default.nix:
```diff
@@ -0,0 +1,29 @@
+{ stdenv, fetchPypi, buildPythonPackage }:
+
+# python3 only
```
`disabled = !isPy3k`

dotlambda commented on pkgs/development/python-modules/aiosmtpd/default.nix:
```diff
@@ -0,0 +1,32 @@
+    atpublic
+  ];
+
+  # error: [Errno 111] Connection refused
```
```suggestion
  # Tests need network access
```
or something along the lines

dotlambda commented on pkgs/development/python-modules/aiosmtpd/default.nix:
```diff
@@ -0,0 +1,32 @@
+{ stdenv, fetchPypi, buildPythonPackage
+, atpublic }:
+
+# python3 only
```
`disabled = !isPy3k`

dotlambda commented on pkgs/development/python-modules/aiosmtpd/default.nix:
```diff
@@ -0,0 +1,32 @@
+  doCheck = false;
+
+  meta = with stdenv.lib; {
+    homepage = "https://aiosmtpd.readthedocs.io/en/latest/";
```
```suggestion
    homepage = https://aiosmtpd.readthedocs.io/en/latest/;
```

dotlambda commented on pkgs/applications/version-management/sr-ht/todo.nix:
```diff
@@ -0,0 +1,39 @@
+with python3Packages;
+
+buildPythonPackage rec {
+  name = "todo-sr-ht-${version}";
```
```suggestion
  pname = "todo-sr-ht";
```

dotlambda commented on pkgs/applications/version-management/sr-ht/todo.nix:
```diff
@@ -0,0 +1,39 @@
+  SRHT_PATH = "${srht}/${python3Packages.python.sitePackages}/srht";
+
+  # Exception: Config option [meta.sr.ht] origin not found
+  doCheck = false;
```
Either fix or explain the error

eadwu replied:
Don't actually know how to fix this but by following the stack trace, it seems like it required a working server? Or at least a configuration file that is present at `/etc/sr.ht/config.ini`.

dotlambda commented on pkgs/applications/version-management/sr-ht/todo.nix:
```diff
@@ -0,0 +1,39 @@
+
+with python3Packages;
+
+buildPythonPackage rec {
```
```suggestion
buildPythonApplication rec {
```

eadwu replied:
Switch the all but `core-sr-ht` to `buildPythonApplication`, since it looks like `core-sr-ht` is suppose to be a python module, `srht`, I'm also guessing I should now move `core-sr-ht` into `pkgs/development/python-modules`?

dotlambda replied:
> I'm also guessing now I should move core-sr-ht into pkgs/development/python-modules?

If it's not going to be used by any other package, you might as well keep it here. But definitely remove it from `all-packages.nix`.

dotlambda commented on pkgs/applications/version-management/sr-ht/todo.nix:
```diff
@@ -0,0 +1,39 @@
+  };
+
+  patches = [
+    ./env.patch
```
Why's this patch needed?

eadwu replied:
I can't seem to get the sitepackages for srht to show unless I explicitly specify `SRHT_PATH` but for it to get to `SRHT_PATH` I need to get rid of the first part in the if statement. https://git.sr.ht/~sircmpwn/git.sr.ht/tree/master/setup.py#L8

PYTHONPATH seems to be fine in the derivation, but `getsitepackages` only returns the python derivation's sitepackages.

dotlambda commented on pkgs/applications/version-management/sr-ht/meta.nix:
```diff
@@ -0,0 +1,44 @@
+
+  meta = with stdenv.lib; {
+    homepage = "https://git.sr.ht/~sircmpwn/meta.sr.ht";
+    description = "meta.sr.ht website";
```
I'm not sure about capitalizing this one.

FRidh replied:
I think if the description would actually describe something the problem wouldn't exist ;-)

dotlambda reviewed (changes requested):
Some requested changes apply to packages that were not yet reviewed as well.

dotlambda commented on pkgs/applications/version-management/sr-ht/core.nix:
```diff
@@ -0,0 +1,70 @@
+  };
+
+  patches = [
+    ./npm.patch
```
What about this one? The files should have more descriptive names.

eadwu replied:
`npm.patch` is suppose to remove the code that calls `npm i` since the derivation doesn't have network access, `node_modules` are fetched through `fetchNodeModules`.

dotlambda commented:
Please squash the changes into the appropriate commits.

eadwu committed:
pythonPackages.pgpy: init at 0.5.2

eadwu committed:
pythonPackages.sqlalchemy-utils: init at 0.34.0

eadwu committed:
python3Packages.asyncpg: init at 0.18.3

eadwu committed:
python3Packages.atpublic: init at 1.0

eadwu committed:
python3Packages.aiosmtpd: init at 1.2

danbst commented on nixos/modules/services/misc/sr-ht.nix:
```diff
@@ -0,0 +1,221 @@
+        type = types.bool;
+        default = false;
+        description = ''
+          Enable the sr-ht service
```
maybe better 
```
Enable sr.ht - git hosting, continuous integration, mailing list, ticket tracking, task dispatching, wiki and account management services.
```

dotlambda commented on pkgs/top-level/all-packages.nix:
```diff
@@ -5581,6 +5581,20 @@ in
 
+  srhtPackages = callPackage ../applications/version-management/sr-ht {
+    # https://github.com/NixOS/nixpkgs/pull/54425#discussion_r249489825
+    pythonPackages = python36Packages;
```
Better use `python36Packages` as argument in the expression and move the comment there.

dotlambda commented on pkgs/top-level/all-packages.nix:
```diff
@@ -5581,6 +5581,20 @@ in
+  srhtPackages = callPackage ../applications/version-management/sr-ht {
+    # https://github.com/NixOS/nixpkgs/pull/54425#discussion_r249489825
+    pythonPackages = python36Packages;
+    fetchNodeModules = callPackage ../applications/networking/instant-messengers/rambox/fetchNodeModules.nix { };
```
This can also be put in a let statement inside the expression. However, I'm not sure if `fetchNodeModules` should be used, I'm not familiar with the Node ecosystem.

dotlambda commented on pkgs/applications/version-management/sr-ht/core.nix:
```diff
@@ -0,0 +1,61 @@
+    pygments
+
+    sassc
+    nodejs-11_x
```
This should not be propagated.

eadwu replied:
Every package that calls `srht` as a dependency then calls its `Makefile` which requires `sassc` and `node` to compile?

dotlambda replied:
No, because there might some package depending on e.g. sr-ht-git but not requiring them. You should put them in each package's `buildInputs`.

eadwu replied:
Is it better to move `sassc` and `nodejs` to `buildInputs` then pass along the `buildInputs` by referencing `buildInputs = srht.buildInputs`?

dotlambda replied:
That's fine as well.

dotlambda commented on pkgs/applications/version-management/sr-ht/git.nix:
```diff
@@ -0,0 +1,44 @@
+, python }:
+
+let
+  redis_2 = redis.overrideAttrs (oldAttrs: rec {
```
```suggestion
  redis_2 = redis.overridePythonAttrs (oldAttrs: rec {
```
You should also use `python.override { packageOverrides = self: super: { ... }; }` and this overriding is probably best done in `default.nix`.

dotlambda commented on pkgs/applications/version-management/sr-ht/git.nix:
```diff
@@ -0,0 +1,44 @@
+
+let
+  redis_2 = redis.overrideAttrs (oldAttrs: rec {
+    inherit (oldAttrs) pname;
```
```suggestion
```

dotlambda commented on pkgs/applications/version-management/sr-ht/git.nix:
```diff
@@ -0,0 +1,44 @@
+    inherit (oldAttrs) pname;
+    version = "2.10.6";
+
+    src = fetchPypi {
```
```suggestion
    src = oldAttrs.src.override {
```

dotlambda commented on pkgs/applications/version-management/sr-ht/git.nix:
```diff
@@ -0,0 +1,44 @@
+    version = "2.10.6";
+
+    src = fetchPypi {
+      inherit pname version;
```
```suggestion
      inherit version;
```

dotlambda commented on pkgs/applications/version-management/sr-ht/disable-npm-install.patch:
```diff
@@ -0,0 +1,15 @@
+-if subprocess.call(["npm", "i"], cwd="srht") != 0:
+-    sys.exit(1)
++# if subprocess.call(["npm", "i"], cwd="srht") != 0:
++#     sys.exit(1)
```
No need to comment the lines out, just delete them.

dotlambda commented on pkgs/development/python-modules/asyncpg/default.nix:
```diff
@@ -0,0 +1,40 @@
+
+  # A useless patch, since flake8 tests are going to fail anyway
+  patches = [
+    ./include-dot-flake8.patch
```
Why did you include it, then? Just get rid of all comments regarding flake8, no need to test for code quality in nixpkgs.

eadwu replied:
Oh, the patch is actually needed for the tests to just not fail, since it first checks if `.flake8` exists before checking whether or not `flake8` is there and skipping if it isn't.

dotlambda commented on pkgs/development/python-modules/atpublic/default.nix:
```diff
@@ -0,0 +1,37 @@
+  checkPhase = ''
+    runHook preCheck
+    pytest --pyargs public
+    runHook postCheck
```
You don't need to include the hooks as none are specified anyway.

dotlambda commented on pkgs/development/python-modules/sqlalchemy-utils/default.nix:
```diff
@@ -0,0 +1,40 @@
+    description = "Various utility functions and datatypes for SQLAlchemy";
+    longDescription = ''
+      SQLAlchemy-Utils provides custom data types and various utility functions
+      for SQLAlchemy.
```
There's really no point in having `longDescription` if all it does is repeating `description`.

dotlambda commented on pkgs/development/python-modules/sqlalchemy-utils/default.nix:
```diff
@@ -0,0 +1,40 @@
+  ];
+
+  # TypeError: sort_query() missing 1 required positional argument: 'query'
+  doCheck = !isPy3k;
```
If I build `python2.pkgs.sqlalchemy-utils`, I get `Ran 0 tests in 0.000s`. A custom `checkPhase` might be needed, probably `py.test sqlalchemy_utils tests`

eadwu replied:
Well quite a lot of tests pass ... while a lot of tests also fail ...

dotlambda commented on pkgs/development/python-modules/weasyprint/default.nix:
```diff
@@ -0,0 +1,47 @@
+    cssselect2
+  ];
+
+  # OSError: cannot load library 'libgobject-2.0.so.0': libgobject-2.0.so.0: cannot open shared object file: No such file or directory
```
You may need to add `gobject-introspection` to `nativeBuildInputs`. I'm not sure if something like `gtk` is needed in `buildInputs` though.

dotlambda reviewed (changes requested):
Please look into how to run pgpy's tests. It's currently giving `Ran 0 tests in 0.000s`.

jb55 commented:
this is awesome

eadwu committed:
sourcehut: init

builds-sr-ht: init at 0.45.13
dispatch-sr-ht: init at 0.11.0
git-sr-ht: init at 0.32.3
hg-sr-ht: init at 0.13.0
lists-sr-ht: init at 0.36.3
man-sr-ht: init at 0.12.4
meta-sr-ht: init at 0.34.3
paste-sr-ht: init at 0.5.1
todo-sr-ht: init at 0.46.8

dotlambda commented on pkgs/development/python-modules/pgpy/default.nix:
```diff
@@ -0,0 +1,49 @@
+  ];
+
+  # Exclude test in root directory from 5 years ago, test_load_asc_bench.py
+  # Has progressbar dependency that doesn't look like even supports Python 3.x
```
There are multiple progressbar successors in our `python-packages.nix`, some of which support Python 2 and 3.

dotlambda replied:
That error should be reported upstream if it has not been fixed yet.

To fix the progressbar2 build, I have pushed 667230441802bcbf278eaf928df28643fbb91ee2 to master.

eadwu replied:
https://github.com/SecurityInnovation/PGPy/issues/250

dotlambda commented on pkgs/development/python-modules/weasyprint/default.nix:
```diff
@@ -0,0 +1,55 @@
+    # tests/test_pdf.py: UnicodeEncodeError: 'ascii' codec can't encode characters in position 65-67: ordinal not in range(128)
+    # tests/test_unicode.py: UnicodeEncodeError: 'ascii' codec can't encode character '\xf8' in position 10: ordinal not in range(128)
+    # tests/test_layout/test_inline.py: AssertionError (assert 100.0 < 100.0)
+    # Other tests just have an x in the output, xfailed
```
xfailed means they are expected to fail. No need to remove them.

dotlambda commented on pkgs/development/python-modules/weasyprint/default.nix:
```diff
@@ -0,0 +1,55 @@
+  ];
+
+  preCheck = ''
+    export LD_LIBRARY_PATH="$LD_LIBRARY_PATH:${lib.makeLibraryPath [ glib pango ]}"
```
Is this only needed for the tests or to run the program in general?

eadwu replied:
`glib` and `pango` are specifically needed for the tests to succeed, though looking at the external dependencies, `cairo` and `gdk_pixbuf` are needed for it to execute.

eadwu committed:
pythonPackages.flake8-polyfill: init at 1.0.2

eadwu committed:
pythonPackages.pep8-naming: init at 0.8.2

eadwu committed:
pythonPackages.pytest-ordering: init at 0.6

dotlambda commented on pkgs/development/python-modules/pgpy/default.nix:
```diff
@@ -0,0 +1,57 @@
+    pytest-ordering
+  ];
+
+  # Exclude broken and outdated test in root directory from 5 years ago, test_load_asc_bench.py
```
That comment seems to be outdated.

dotlambda commented on pkgs/development/python-modules/pgpy/default.nix:
```diff
@@ -0,0 +1,57 @@
+  version = "0.4.3";
+
+  patches = [
+    ./disable-failing-tests.patch
```
No idea about why they're failing. Might be worth reporting upstream.

eadwu replied:
Expanded https://github.com/SecurityInnovation/PGPy/issues/250.

dotlambda commented on nixos/modules/services/misc/sr-ht.nix:
```diff
@@ -0,0 +1,771 @@
+    domain=man.sr.ht.local
+    #
+    # A secret key to encrypt session cookies with.
+    secret-key=CHANGEME
```
Shouldn't that be configurable?

eadwu replied:
Yeah, working on the module still.

dotlambda commented on nixos/modules/services/misc/sr-ht.nix:
```diff
@@ -0,0 +1,771 @@
+    ${optionalString (elem "git" cfg.services) "git=http://git.sr.ht.local"}
+    ${optionalString (elem "builds" cfg.services) "builds=http://builds.sr.ht.local"}
+    ${optionalString (elem "todo" cfg.services) "todo=http://todo.sr.ht.local"}
+    ${optionalString (elem "man" cfg.services) "man=http://man.sr.ht.local"}
```
As well as these URLs?

dotlambda commented on nixos/modules/services/misc/sr-ht.nix:
```diff
@@ -0,0 +1,771 @@
+    # command and do not enter a new password.
+    pgp-privkey=
+    pgp-pubkey=
+    pgp-key-id=
```
As well as all these unset options?

eadwu replied:
By default they're unset, don't actually know if they are required yet.

dotlambda reviewed (commented):
The amount of options that would be needed seems way too big for all to be included in the NixOS module. How about just letting the user specify the whole configuration file?

eadwu commented:
I was planning on splitting the modules into different files `srht/builds.nix`, `srht/dispatch.nix`, etc and having `extraConfig` in each one with all the modules imported from `srht/default.nix`.

nixos-discourse commented:
This pull request has been mentioned on **Nix community**. There might be relevant details there:

https://discourse.nixos.org/t/builds-sr-ht-can-now-run-builds-in-a-nixos-image/1924/6

eadwu commented:
Running to an error from `gunicorn` trying to start `meta.sr.ht`, namely `ModuleNotFoundError: No module named 'email.mime'; 'email' is not a package`. Trying `import email` in `python3` works just fine.

adisbladis commented on nixos/modules/services/misc/sr-ht/builds.nix:
```diff
@@ -0,0 +1,202 @@
+        '';
+      };
+
+      worker.user = mkOption {
```
Does the user need to be a "real" user or would a systemd `DynamicUser` be fine?

eadwu replied:
Just did a bit of reading and using `DynamicUser` might be possible if the logs go into `/var` or `/var/tmp` and if I can serve the directory through nginx, since I don't know if it's possible to access it with `PrivateTmp` on.

arianvp replied:
Do note that `DynamicUser` is currently broken in `18.09`   (I fixed it in `master` and `18.09`)

eadwu replied:
I don't there are any plans to backport this anyway, seeing how `19.03` is coming soon.

eadwu commented on pkgs/applications/version-management/sr-ht/core.nix:
```diff
@@ -0,0 +1,75 @@
+  '';
+
+  # No actual? tests but seems like it needs this anyway
+  preCheck = let
```
I have no clue what's going on here but the `checkPhase` succeeds now so that's good?

eadwu commented on nixos/modules/services/misc/sr-ht/meta.nix:
```diff
@@ -0,0 +1,274 @@
+
+      description = "meta.sr.ht website service";
+      path = with pkgs; [ python36Packages.gunicorn ];
+      environment.PYTHONPATH = builtins.concatStringsSep ":" (with pkgs.python36Packages; [
```
I'm pretty certain this isn't the way I should be doing this, including `meta-sr-ht` in `path` only includes the `/bin` directory.

eadwu replied:
https://github.com/NixOS/nixpkgs/pull/54425#discussion_r250533104

dotlambda commented on nixos/modules/services/misc/sr-ht/git.nix:
```diff
@@ -0,0 +1,115 @@
+      wantedBy = [ "multi-user.target" ];
+
+      description = "git.sr.ht website service";
+      path = with pkgs; [ (python36Packages.gunicorn.overridePythonAttrs (oldAttrs: {
```
It should be sufficient to use `python36.withPackages`.

eadwu replied:
`python36.withPackages` doesn't seem to work with the `*-sr-ht` packages.

dotlambda replied:
Please be more specific.

dotlambda replied:
Makes sense since meta-sr-ht is using `buildPythonApplication`.

dotlambda commented on pkgs/development/python-modules/weasyprint/default.nix:
```diff
@@ -0,0 +1,61 @@
+
+  postInstall = ''
+    wrapProgram $out/bin/weasyprint \
+      --set LD_LIBRARY_PATH $LD_LIBRARY_PATH
```
That does not work if it's imported. You should instead use a patch and `substituteAll` to hardcode the paths here: https://github.com/Kozea/WeasyPrint/blob/v44/weasyprint/text.py#L235-L240.

eadwu replied:
Should `libfontconfig.so.1` also be patch or is it safe to assume that `fontconfig` is always in `LD_LIBRARY_PATH`?

eadwu replied:
Oh, tests broke due to line length, is it better to patch in an increase to `max-line-length` or just disable `flake8` tests altogether.

dotlambda replied:
If you're at it, yes, fonconfig should be patched in as well.

There's no value in running flake8 tests and you can hence safely disable them. However, that might be more work than just adding an exception for the respective lines. See http://flake8.pycqa.org/en/3.1.1/user/ignoring-errors.html#in-line-ignoring-errors.

dotlambda commented on pkgs/applications/version-management/sr-ht/default.nix:
```diff
@@ -0,0 +1,31 @@
+# https://github.com/NixOS/nixpkgs/pull/54425#discussion_r249489825
+
+let
+  fakePython = python36.override {
```
```suggestion
  python = python36.override {
```

dotlambda commented on pkgs/applications/version-management/sr-ht/default.nix:
```diff
@@ -0,0 +1,31 @@
+          sha256 = "03vcgklykny0g0wpvqmy8p6azi2s078317wgb2xjv5m2rs9sjb52";
+        };
+      });
+    };
```
```suggestion
    };
    core-sr-ht = self.callPackage ./core.nix { inherit fetchNodeModules; };
    meta-sr-ht = self.callPackage ./meta.nix { };
```

eadwu replied:
If I interpret this correctly, switch `buildPythonApplication` to `buildPythonPackage`, find a way to index the new `python` with the modules and then use `withPackages` in it?

dotlambda replied:
Exactly. And you can make `python` know about the modules by putting them in `packageOverrides`.

dotlambda replied:
All of them? I had the impression only core and meta would ever be used as modules. And no, I really don't think there's any use in having them in `python-packages.nix`.

eadwu replied:
Yeah, since it seems that when `gunicorn` is run, it imports the module from each package.

dotlambda replied:
Ok, then just use `buildPythonPackage` and add a comment on why it's needed.

dotlambda commented on pkgs/applications/version-management/sr-ht/default.nix:
```diff
@@ -0,0 +1,31 @@
+  git = fakePython.pkgs.callPackage ./git.nix { inherit srht; };
+  lists = fakePython.pkgs.callPackage ./lists.nix { inherit srht; };
+  man = fakePython.pkgs.callPackage ./man.nix { inherit srht; };
+  meta = fakePython.pkgs.callPackage ./meta.nix { inherit srht; };
```
```suggestion
  meta = with python.pkgs; toPythonApplication meta-sr-ht;
```

dotlambda commented on pkgs/development/python-modules/weasyprint/default.nix:
```diff
@@ -0,0 +1,68 @@
+    substituteInPlace weasyprint/text.py \
+      --subst-var-by libgobject-2-0 ${lib.makeLibraryPath [ glib ]}/libgobject-2.0.so \
+      --subst-var-by libpango-1-0 ${lib.makeLibraryPath [ pango ]}/libpango-1.0.so \
+      --subst-var-by libpangocairo-1-0 ${lib.makeLibraryPath [ pango ]}/libpangocairo-1.0.so
```
Why not use `substituteAll`? That's at least what I usually do but it's up to personal preference.

dotlambda replied:
No, instead of using `substituteInPlace` you can use `substituteAll` inside `patches`. Just search for it and you'll see how it works.
Btw, you can e.g. put `libgobject-2-0 = "${lib.makeLibraryPath [ glib ]}/libgobject-2.0.${stdenv.hostPlatform.extensions.sharedLibrary}"` in there.

eadwu replied:
Oh, I only knew about the `substituteAll` as a function `substituteAll inFile outFile` from nixpkgs/manual, and turns out there is a equivalent of `makeLibraryPath` for one package, `getLib`, so I'll change that to.

dotlambda commented on pkgs/development/python-modules/weasyprint/ld-library-path.patch:
```diff
@@ -0,0 +1,36 @@
++gobject = dlopen(ffi, 'gobject-2.0', 'libgobject-2.0-0', '@libgobject-2-0@', # noqa: E261,E501
+                  'libgobject-2.0.dylib')
+-pango = dlopen(ffi, 'pango-1.0', 'libpango-1.0-0', 'libpango-1.0.so',
++pango = dlopen(ffi, 'pango-1.0', 'libpango-1.0-0', '@libpango-1-0@', # noqa: E261,E501
```
```suggestion
+pango = dlopen(ffi, '@libpango-1-0@') # noqa: E261,E501
```
No need to have multiple options if we already know the one.

dotlambda commented on pkgs/development/python-modules/weasyprint/default.nix:
```diff
@@ -0,0 +1,68 @@
+
+  postPatch = ''
+    substituteInPlace weasyprint/fonts.py \
+      --subst-var-by libpangoft2-1-0 ${lib.makeLibraryPath [ pango ]}/libpangoft2-1.0.so \
```
```suggestion
      --subst-var-by libpangoft2-1-0 ${lib.makeLibraryPath [ pango ]}/libpangoft2-1.0.${stdenv.hostPlatform.extensions.sharedLibrary} \
```
Not sure about the advantage of using `makeLibraryPath`.

eadwu replied:
It seems better to use for multiple outputs, like for `fontconfig` at least, since `lib/` is within `fontconfig.lib`.

colemickens commented on nixos/modules/services/misc/sr-ht/builds.nix:
```diff
@@ -0,0 +1,183 @@
+
+          For more information about the security model of builds.sr.ht, visit the wiki:
+
+          https://man.sr.ht/builds.sr.ht/security
```
This URL doesn't exist?

eadwu replied:
Looks like the example config is outdated in some parts, looks like it's now in with the installation instructions? https://man.sr.ht/builds.sr.ht/installation.md#security-model

tilpner replied:
Perhaps this was meant to be https://man.sr.ht/builds.sr.ht/installation.md#security-model ?

tilpner commented:
[This should probably be renamed now.](https://sourcehut.org/)

eadwu commented:
Yeah, just saw it at here [1]. Looks like the default configs were updated too, planning on updating everything over the weekend and finishing this up by the end of mid-winter break.

[1] https://news.ycombinator.com/item?id=19100986

eadwu commented on pkgs/applications/version-management/sourcehut/core.nix:
```diff
@@ -0,0 +1,75 @@
+    '';
+  in ''
+    # Tests need config option "origin" from "meta.sr.ht"
+    cp ${config} config.ini
```
Not really someone who actively uses Python, but looking at the code [1], there is a possibility that the local `config.ini` overrides the global configuration. However, I don't see any `config.ini`s (find -name '*.ini') in the output derivation so I believe it shouldn't matter.

The login/register links in meta.sr.ht work fine for me, however, there is a "broken" link that leads to `sr.ht` instead of `/login`

[1] https://git.sr.ht/~sircmpwn/core.sr.ht/tree/master/srht/config.py#L12

tomberek replied:
I think I was able to resolve this with service.srht.*.domain settings.

tomberek commented:
Status? This still looks good. Do we need more review?

eadwu commented:
Don't really have time to work on this anytime soon. The packaging is more or less fine (if I have time I'll update the versions later today) but the module for the services is still incomplete, or rather they are hosted successfully but don't actually have the functionality they're suppose to have.

zimbatm commented:
this just needs one last rebase to fix the merge conflict. looks good to go otherwise.

eadwu commented:
Pushed to fix conflict, build has problems right now, among which `pgpy` has a broken test, `cairosvg` has dependency issues, and `*srht` also issues since it can't seem to find `sassc` from `buildInputs`.

tomberek commented:
The CairoSVG test failure is being addressed upstream: https://github.com/tiran/defusedxml/pull/24

eadwu commented:
Depends on #56690 and #59425. Updated to latest versions and looks like I'll have to wait for more packages to sync to pypi or upstream fixes.

eadwu commented:
Builds fine now, waiting on #56690.

eadwu commented on pkgs/applications/version-management/sourcehut/default.nix:
```diff
@@ -0,0 +1,47 @@
+      };
+
+      pyopenssl = super.pyopenssl.override {
+        openssl = openssl_1_1;
```
Don't know whether or not this should be handled on `python-packages.nix`.

baracoder commented:
@eadwu #56690 is merged now

eadwu commented:
Need to wait for staging to merge to master, then I'll update the packages again to see if anything breaks.

eadwu commented:
Waiting on #61305.

eadwu commented:
#61679.

eadwu commented:
Everything builds fine now.

matthiasbeyer commented:
Any progress here?

eadwu committed:
python3Packages.emailthreads: init at 0.1.0

eadwu commented:
Everything builds fine as for the module, I haven't worked on it in a while.

matthiasbeyer commented:
And there are no pending questions as far as I can see ... I guess this can be merged then, right?

matthiasbeyer commented:
As far as I can see, this is only package definitions, right? I would opt for merging this rather sooner than later and then get started working on services (and tests for them!!!).

Anything I can help with to get this merged ASAP?

@dotlambda @eadwu ?

joachifm commented:
@GrahamcOfBorg build sourcehut

joachifm commented:
ofborg reports build failure, is that expected?

eadwu commented:
Rebased onto the latest master and it still failed, looks like it came from `pytest_3`. Should be fixed now.

joachifm commented:
@GrahamcOfBorg build sourcehut

matthiasbeyer commented:
Still no progress here?

eadwu commented:
I'm intermittently working on the modules still since I'm planning on hosting it in college but on the package drv side it should be all good.

eadwu commented:
The status update that shows a [somewhat] working `sr.ht` finally. Really only tested `meta`, `git`, and `paste` while `todo`, `man`, and `hg` works fine [so far] from my tinkering. Local nixpkgs consist of https://github.com/NixOS/nixpkgs/pull/66401, https://github.com/NixOS/nixpkgs/pull/65109, and https://github.com/NixOS/nixpkgs/pull/54425 on top of 2da00ae7b4942e3946041d1e0b19120a457d33e3 with a nixops configuration here [1].

In terms of other stuff, the mailing/mail server will probably be the last thing that'll work since I hardly have any experience setting that up. I'll probably just setup the web interfaces for `dispatch`, `lists`, and `builds` while focusing on `builds` so it'll be completed before I head for the dorms in college.

[1] https://gist.github.com/eadwu/e633df2660b316f3d09974594cb519e5

bsima commented:
Edmund, for the mailserver, it might help to look at
https://gitlab.com/simple-nixos-mailserver/nixos-mailserver for reference...

Then again, setting up a mailserver could be outside the scope of this,
since it is a dependency of sourcehut and not part of the software suite itself.

matthiasbeyer commented:
> Then again, setting up a mailserver could be outside the scope of this,

I agree. I also suggest the nixos-mailserver, which I have good experience with (used it only for journalwatch so far).
I would also say that this is out of scope of this package and a user should setup their own webserver for using sourcehut, not the package/service should ship one!

joachifm commented on pkgs/development/python-modules/asyncpg/aaeb707.patch:
```diff
@@ -0,0 +1,37 @@
+From aaeb7076e5acb045880b46155014c0640624797e Mon Sep 17 00:00:00 2001
```
I know this is a small patch but is there any way to get it from an external source?

eadwu replied:
Now retrieved from https://github.com/MagicStack/asyncpg/commit/aaeb7076e5acb045880b46155014c0640624797e.patch.

joachifm commented:
Seems to me all review items have been addressed. Unless there is something fundamentally wrong with this, it'd be nice to finally integrate it.

eadwu commented:
Rebased onto latest master and built successfully.

joachifm commented:
Thank you

tomberek commented:
Bravo! @eadwu thanks for sticking with it.

matthiasbeyer commented:
Yippie!

aaronjanse commented:
Thanks @eadwu! This closes #65109, right?

eadwu commented:
#65109 is an incomplete implementation of an actual sourcehut network (internals, not webserver/mailserver). This PR just included the packages needed for it to function.

pinpox commented:
Is it possible to run sourcehut with these packages and systemd services until the module in #65109 is merged?
This is what I am trying, but the service doesn't find the dependencies:

```nix
  systemd.services.sourcehut-meta = {
      description = "meta.sr.ht website service";
      requires = ["postgresql.service"];
      after = [ "postgresql.service network.target" ];

      environment = let
        penv = pkgs.python.buildEnv.override {
          extraLibs = [ pkgs.sourcehut.metasrht ] ;
        };
      in
      {
              PYTHONPATH = "${penv}/${pkgs.python.sitePackages}";
      };

      serviceConfig = {
        Type = "simple";
        User = "sourcehut";
        Restart = "always";
        ExecStart = ''${pkgs.python38Packages.gunicorn}/bin/gunicorn app.py --chdir ${pkgs.sourcehut.metasrht}/lib/python3.8/site-packages/metasrht -b 127.0.0.1:5000'';
      };
   };
```

eadwu commented:
`meta.sr.ht` was implemented to a fully functional state already when the PR was created.

pinpox commented:
> `meta.sr.ht` was implemented to a fully functional state already when the PR was created.

Can you link me any kind of information or docs on how to actually run it?

eadwu commented:
I believe a nixops implementation was referred on IRC? And there is the nixops implementation I used locally within nixos-configuration, although it's now moved from `sr.ht` to `github`.

There's no difference between the nixops configuration and normal NixOS configuration (just combine it all), nixops is just a easier way to distribute sourcehut across multiple machines (VMs).

Wrong thread to post on though.

sumnerevans commented on pkgs/applications/version-management/sourcehut/default.nix:
```diff
@@ -0,0 +1,49 @@
+
+  python = python37.override {
+    packageOverrides = self: super: {
+      srht = self.callPackage ./core.nix { inherit fetchNodeModules; };
```
Is there a way to expose this package within a nix-shell? (Context: trying to do development for builds.sr.ht on NixOS.)

sumnerevans replied:
@eadwu since you are listed as the maintainer

eadwu replied:
`sourcehut.python.pkgs.srht`

pinpox replied:
In case you switch threads, please mention me. I'm trying to keep an eye on the progress of this.

sumnerevans replied:
Oh, wow, I feel really dumb now. I missed that `python` was inherited. Thanks for the quick response!
